zebra/zebra_mlag.c

```diff
@@ -23,16 +23,559 @@
+		 * write to MCLAGD
+		 */
+		if (len > 0)
+			zebra_mlag_private_write_data(mlag_wr_buffer, len);
+
+		/*
+		 * If mesasge type is De-register, send a signal to main thread,
+		 * sothat necessary cleanup will be done by main thread.
+		 */
+		if (msg_type == MLAG_DEREGISTER) {
+			thread_add_event(zrouter.master,
+					 zebra_mlag_terminate_pthread, NULL, 0,
+					 NULL);
+		}
+
+		stream_free(s);
+	}
+
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug(":%s: Posted  %d messages to MLAGD", __func__,
+			   wr_count);
+	/*
+	 * Currently there is only message write task is enqueued to this
+	 * thread, yielding was added for future purpose, sothat this thread can
+	 * server other tasks also and in case FIFO is empty, this task will be
+	 * schedule when main thread adds some messages
+	 */
+	if (wr_count >= ZEBRA_MLAG_POST_LIMIT)
+		zebra_mlag_signal_write_thread();
+	return 0;
+}
+
+/*
+ * API to handle teh process state.
+ * In case of Down, Zebra keep monitoring the MLAG state.
+ * all the state Notifications will be published to clients
+ */
+void zebra_mlag_handle_process_state(enum zebra_mlag_state state)
+{
+	if (state == MLAG_UP) {
+		zrouter.mlag_info.connected = true;
+		zebra_mlag_publish_process_state(NULL, ZEBRA_MLAG_PROCESS_UP);
+		zebra_mlag_send_register();
+	} else if (state == MLAG_DOWN) {
+		zrouter.mlag_info.connected = false;
+		zebra_mlag_publish_process_state(NULL, ZEBRA_MLAG_PROCESS_DOWN);
+		zebra_mlag_private_monitor_state();
+	}
+}
+
+/***********************End of MLAG Thread processing*************************/
+
+/*************************Multi-entratnt Api's********************************/
+
+/*
+ * Provider api to signal that work/events are available
+ * for the Zebra MLAG Write pthread.
+ * This API is called from 2 pthreads..
+ * 1) by main thread when client posts a MLAG Message
+ * 2) by MLAG Thread, in case of yield
+ * though this api, is called from two threads we don't need any locking
+ * because Thread task enqueue is thread safe means internally it had
+ * necessary protection
+ */
+static int zebra_mlag_signal_write_thread(void)
+{
+	if (zrouter.mlag_info.zebra_pth_mlag) {
+		if (IS_ZEBRA_DEBUG_MLAG)
+			zlog_debug(":%s: Scheduling MLAG write", __func__);
+		thread_add_event(zrouter.mlag_info.th_master,
+				 zebra_mlag_thread_handler, NULL, 0,
+				 &zrouter.mlag_info.t_write);
+	}
+	return 0;
+}
+
+/*
+ * API will be used to publish the MLAG state to interested clients
+ * In case client is passed, state is posted only for that client,
+ * otherwise to all interested clients
+ * this api can be called from two threads.
+ * 1) from main thread: when client is passed
+ * 2) from MLAG Thread: when client is NULL
+ *
+ * In second case, to avoid global data access data will be post to Main
+ * thread, so that actual posting to cleints will happen from Main thread.
+ */
+static void zebra_mlag_publish_process_state(struct zserv *client,
+					     zebra_message_types_t msg_type)
+{
+	struct stream *s = NULL;
```

Comment:
Don't set this to NULL so the compiler can catch cases where this is used uninitialized.